File: rightsideview.py
# leetcode:accepted
# time complexity: O(n)
# space complexity: O(n)
# explaination:
"""
Two methods - BFS and DFS
1) in bfs, you do level ordrer traversal and push the elements into the queue and keep a track of the size of the q. you iterate through all the elements of a level that is pushed in the q, and keep a variable size wich keeps track of size of nodes at a particular level. When your i reaches size-1, then you will add the element to the resultant list

2) In DFS method, you keep on iterating towards the right side of the tree. You add the value to the resultant list when the depth of the tree is equal to the size of the tree


"""

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

#BFS method
class Solution:
    def rightSideView(self, root: TreeNode) -> List[int]:
        queue=[]
        output=[]
        if root==None:
            return output
        queue.append(root)  #ppending the root to the queue
        while(len(queue)!=0):  #you keep on processinguntil ur queue is empty
            size=len(queue)
            for i in range(len(queue)):
                element=queue.pop(0)
                if i==size-1 :  #pushing the last element at each level
                    output.append(element.val)
                if element.left!=None:
                    queue.append(element.left)  #appending the left child
                if element.right!=None:
                    queue.append(element.right)    #appending the right child
        return output

    #BFS almost same method
from collections import deque
class Solution(object):
    def rightSideView(self, root):

        ##BFS
        q=deque()
        result=[]
        if not root:
            return None
        q.append(root)
        while(len(q)!=0):
            size=len(q)
            temp=None
            for i in range(size):
                item=q.popleft()
                value=item.val
                if item.left!=None:
                    q.append(item.left)
                if item.right!=None:
                    q.append(item.right)
            result.append(value)  #you just add the value to the result when your i is terminating
        return result



class Solution:
    def rightSideView(self, root: TreeNode) -> List[int]:
        result = []
        self.helper(root, result, 0)
        return result

    def helper(self, root, result, depth):
        if not root:
            return None
        if len(result) == depth:
            result.append(root.val)
        self.helper(root.right, result, depth + 1)
        self.helper(root.left, result, depth + 1)

File: right_side_view.cpp
//Time Complexity-O(n)
//Space complexity-O(n/2)=O(n)
//Ran successfully on leetcode
//3 point algo-
	//1a)In DFS, just process the right side first
	//1b)If the size of the result vector is the same as level, then we add it to the result vector. That makes it the right most
	//2a)In bfs, iterate for a particular level by kepping track using the queue
	//2b)We track the last element in a particular level. Last always gets overwritten by the last node we traverse in that level
	//2c)Pop the node whose children have been added

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
    vector<int>result;
public:
    /*vector<int> rightSideView(TreeNode* root) {
        if(root==NULL)
            return {};
        dfs(root,0);
        return result;
    }
    void dfs(TreeNode* root,int level)
    {
        if(root==NULL)
            return;
        if(level==result.size())
            result.push_back(root->val);
        dfs(root->right,level+1);
        dfs(root->left,level+1);
    }*/
    //bfs method
    vector<int> rightSideView(TreeNode* root) {
        if(root==NULL)
            return {};
        queue<TreeNode*>q;
        int last;
        q.push(root);
        while(!q.empty())
        {
            int n=q.size();
            for(int i=0;i<n;i++)
            {
                TreeNode*curr=q.front();
                last=curr->val;
                //add children of the node
                if(curr->left!=NULL)
                    q.push(curr->left);
                if(curr->right!=NULL)
                    q.push(curr->right);
                q.pop();
            }
            result.push_back(last);
        }
        return result;
    }
};